src/cryptography.ts

```diff
@@ -0,0 +1,176 @@
+   * @param iterations - The number of iterations.
+   * @param keyLength - The desired key length.
+   * @returns The PBKDF2 of the password.
+   */
+  pbkdf2Sha512?: (
+    password: Uint8Array,
+    salt: Uint8Array,
+    iterations: number,
+    keyLength: number,
+  ) => Promise<Uint8Array>;
+};
+
+/**
+ * Compute the HMAC-SHA-512 of the given data using the given key.
+ *
+ * This function uses the Web Crypto API if available, falling back to a
+ * JavaScript implementation if not.
+ *
+ * @param key - The key to use.
+ * @param data - The data to hash.
+ * @param cryptographicFunctions - The cryptographic functions to use. If
+ * provided, these will be used instead of the built-in implementations.
+ * @returns The HMAC-SHA-512 of the data.
+ */
+export async function hmacSha512(
+  key: Uint8Array,
+  data: Uint8Array,
+  cryptographicFunctions: CryptographicFunctions = {},
+): Promise<Uint8Array> {
+  if (cryptographicFunctions.hmacSha512) {
+    return await cryptographicFunctions.hmacSha512(key, data);
+  }
+
+  if (isWebCryptoSupported()) {
+    /* eslint-disable no-restricted-globals */
+    const subtleKey = await crypto.subtle.importKey(
+      'raw',
+      key,
+      { name: 'HMAC', hash: 'SHA-512' },
+      false,
+      ['sign'],
+    );
+
+    const result = await crypto.subtle.sign('HMAC', subtleKey, data);
+    return new Uint8Array(result);
+    /* eslint-enable no-restricted-globals */
+  }
+
+  return nobleHmac(nobleSha512, key, data);
+}
+
+/**
+ * Compute the Keccak-256 of the given data synchronously.
+ *
+ * Right now this is just a wrapper around `keccak256` from the `@noble/hashes`
+ * package, but it's here in case we want to change the implementation in the
+ * future to allow for asynchronous hashing.
+ *
+ * @param data - The data to hash.
+ * @returns The Keccak-256 of the data.
+ */
+export function keccak256(data: Uint8Array): Uint8Array {
+  return nobleKeccak256(data);
+}
+
+/**
+ * Compute the PBKDF2 of the given password, salt, iterations, and key length.
+ * The hash function used is SHA-512.
+ *
+ * @param password - The password to hash.
+ * @param salt - The salt to use.
+ * @param iterations - The number of iterations.
+ * @param keyLength - The desired key length.
+ * @param cryptographicFunctions - The cryptographic functions to use. If
+ * provided, these will be used instead of the built-in implementations.
+ * @returns The PBKDF2 of the password.
+ */
+export async function pbkdf2Sha512(
+  password: Uint8Array,
+  salt: Uint8Array,
+  iterations: number,
+  keyLength: number,
+  cryptographicFunctions: CryptographicFunctions = {},
+): Promise<Uint8Array> {
+  if (cryptographicFunctions.pbkdf2Sha512) {
+    return await cryptographicFunctions.pbkdf2Sha512(
+      password,
+      salt,
+      iterations,
+      keyLength,
+    );
+  }
+
+  if (isWebCryptoSupported()) {
+    /* eslint-disable no-restricted-globals */
+    const result = await crypto.subtle.importKey(
```

Comment:
If possible we should test the two implementations against each other and compare the result

Reply:
There's a test that uses the Web Crypto API and one that doesn't, and we use the same inputs for both. Is that not sufficient?